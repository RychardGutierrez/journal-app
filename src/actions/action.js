export const newAction = (type, payload) => ({ type, payload });
