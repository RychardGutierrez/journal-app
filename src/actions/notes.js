import { deleteDoc, doc, updateDoc } from 'firebase/firestore';
import {
  fireStoreAddDoc,
  fireStoreCollection,
  fireStoreDataBase,
} from '../firebase/firebase-config';

import { NOTE_PATH, NEW_NOTE } from '../common/constants';
import { loadNotes } from '../common/loadNotes';
import { imgFileUpload } from '../common/utils';
import { types } from '../types';
import { newAction } from './action';
import {
  messageClose,
  messageNoteDeleteWait,
  messageNoteUpdate,
} from '../common/messages';
import { messageNoteUpdateError } from './../common/errorMessage';
import { messageNoteUpdateWait } from './../common/messages';

export const startNewNote = () => async (dispatch, getState) => {
  try {
    const { uid } = getState().auth;
    const documentRef = await fireStoreAddDoc(
      fireStoreCollection(fireStoreDataBase, `${uid}${NOTE_PATH}`),
      NEW_NOTE
    );

    dispatch(activeNote(documentRef.id, NEW_NOTE));
    dispatch(startLoadingNotes(uid));
  } catch (error) {
    console.log(error);
  }
};

export const activeNote = (id, note) =>
  newAction(types.notesActive, { ...note, id });

export const startLoadingNotes = (uid) => async (dispatch) => {
  const notes = await loadNotes(uid);
  dispatch(newAction(types.notesLoad, notes));
};

export const startLoadingEntry = (id) => (dispatch, getState) => {
  const { data } = getState().notes;
  dispatch(activeNote(id, data[id]));
};

export const refreshNote = (id, note) =>
  newAction(types.notesUpdated, { ...note, id });

export const startSaveNote = (note) => async (dispatch, getState) => {
  try {
    const { uid } = getState().auth;
    const { active } = getState().notes;
    const noteReference = doc(
      fireStoreDataBase,
      `${uid}${NOTE_PATH}/${active.id}`
    );
    const noteToFireStore = { ...note };

    !noteToFireStore.url && delete noteToFireStore.url;
    delete noteToFireStore.id;
    await updateDoc(noteReference, noteToFireStore);
    dispatch(refreshNote(note.id, noteToFireStore));

    messageNoteUpdate();
  } catch (error) {
    messageNoteUpdateError();
  }
};

export const startUploadingImg = (imgFile) => async (dispatch, getState) => {
  messageNoteUpdateWait();

  const { active } = getState().notes;
  const fileUrl = await imgFileUpload(imgFile);
  const newNoteActive = { ...active, url: fileUrl };
  dispatch(startSaveNote(newNoteActive));

  messageClose();
};

const deleteNote = (id) => newAction(types.notesDelete, { id });

export const startDeleteNote = (id) => async (dispatch, getState) => {
  try {
    messageNoteDeleteWait();

    const { uid } = getState().auth;
    const noteRef = doc(fireStoreDataBase, `${uid}${NOTE_PATH}/${id}`);
    await deleteDoc(noteRef);
    dispatch(deleteNote(id));

    messageClose();
  } catch (error) {
    console.log(error);
  }
};

export const cleanNote = () => newAction(types.notesLogoutClean);
