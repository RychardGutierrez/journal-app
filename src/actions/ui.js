import { types } from "../types";
import { newAction } from "./action";

export const uiSetErrorMessage = ({ messageError }) =>
  newAction(types.uiSetError, { messageError });

export const uiRemoveErrorMessage = () => newAction(types.uiRemoveError);

export const uiStartLoading = () => newAction(types.uiStartLoading);

export const uiEndLoading = () => newAction(types.uiEndLoading);
