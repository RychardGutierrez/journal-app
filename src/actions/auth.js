import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword as createUserFirebase,
  updateProfile,
  signInWithEmailAndPassword as signInUserFirebase,
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';

import { types } from '../types';
import { newAction } from './action';
import { cleanNote } from './notes';
import { uiEndLoading, uiSetErrorMessage, uiStartLoading } from './ui';
import {
  messageErrorRegisterNewUser,
  messageLoginError,
} from '../common/errorMessage';
import { messageWelcome } from '../common/messages';

export const login = (uid, name) => newAction(types.login, { uid, name });

export const startLogout = () => async (dispatch) => {
  try {
    await getAuth().signOut();
    dispatch(newAction(types.logout));
    dispatch(cleanNote());
  } catch (error) {
    console.log(error);
  }
};

export const startLoginEmailPassword =
  ({ email, password }) =>
  async (dispatch) => {
    dispatch(uiStartLoading());
    try {
      const { user } = await signInUserFirebase(getAuth(), email, password);
      const response = dispatch(login(user.uid, user.displayName));

      messageWelcome(user.displayName);
      return response;
    } catch (error) {
      return messageLoginError();
    } finally {
      return dispatch(uiEndLoading());
    }
  };

export const startGoogleLogin = () => async (dispatch) => {
  try {
    const { user } = await signInWithPopup(getAuth(), googleAuthProvider);
    dispatch(login(user.uid, user.displayName));

    messageWelcome(user.displayName);
  } catch (error) {
    messageLoginError();
  } finally {
    dispatch(uiEndLoading());
  }
};

export const startRegisterWithEmailPasswordName =
  ({ email, password, name }) =>
  async (dispatch) => {
    dispatch(uiStartLoading());
    try {
      const { user } = await createUserFirebase(getAuth(), email, password);
      await updateProfile(user, { displayName: name });
      dispatch(login(user.uid, user.displayName));

      messageWelcome(user.displayName);
    } catch (error) {
      dispatch(uiSetErrorMessage(messageErrorRegisterNewUser));
    } finally {
      dispatch(uiEndLoading());
    }
  };
