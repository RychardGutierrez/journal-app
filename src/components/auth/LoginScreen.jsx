import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import isEmail from 'validator/lib/isEmail';
import isStrongPassword from 'validator/lib/isStrongPassword';

import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { uiRemoveErrorMessage, uiSetErrorMessage } from '../../actions/ui';
import { messageErrorWrongEmailOrPassword } from '../../common/errorMessage';
import { LOGIN, SIGN_GOOGLE, CREATE_NEW_ACCOUNT } from '../../common/constants';

const LoginScreen = () => {
  const dispatch = useDispatch();
  const { messageError, loading } = useSelector(({ ui }) => ui);

  // TODO remove this init state
  // email: 'lucas@gmail.com',
  // password: '123Lucas@',
  const { values, handleInputChange } = useForm({
    email: '',
    password: '',
  });

  const { email, password } = values;

  const handleLogin = (e) => {
    e.preventDefault();
    isFormValid() && dispatch(startLoginEmailPassword({ email, password }));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  const isFormValid = () => {
    if (!isEmail(email)) {
      dispatch(uiSetErrorMessage(messageErrorWrongEmailOrPassword));
      return false;
    }

    if (password.trim().length === 0 || !isStrongPassword(password)) {
      dispatch(uiSetErrorMessage(messageErrorWrongEmailOrPassword));
      return false;
    }

    messageError && dispatch(uiRemoveErrorMessage());
    return true;
  };

  return (
    <>
      <h3 className="auth__title">{LOGIN}</h3>
      <form
        onSubmit={handleLogin}
        className="animate__animated animate__fadeIn animate_faster"
      >
        {messageError && (
          <div className="auth__alert-error">{messageError}</div>
        )}

        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          {loading ? 'Loading...' : 'Loading'}
        </button>
        <hr />
        <div className="auth__social_networks ">
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>{SIGN_GOOGLE}</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register" className="link">
          {CREATE_NEW_ACCOUNT}
        </Link>
      </form>
    </>
  );
};

export default LoginScreen;
