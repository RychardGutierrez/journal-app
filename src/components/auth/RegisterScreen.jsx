import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

import isEmail from 'validator/lib/isEmail';
import isStrongPassword from 'validator/lib/isStrongPassword';

import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { uiRemoveErrorMessage, uiSetErrorMessage } from '../../actions/ui';
import { ALREADY_REGISTER, REGISTER } from '../../common/constants';
import {
  messageErrorEmail,
  messageErrorName,
  messageErrorPassword,
} from '../../common/errorMessage';
import { useForm } from '../../hooks/useForm';

const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { messageError } = useSelector(({ ui }) => ui);

  // TODO remove this init state
  // name: "Lucas the duck",
  // email: "lucas.duck@gmail.com",
  // password: "123Lucas@",
  // confirmPassword: "123Lucas@",

  const { values, handleInputChange } = useForm({
    name: '',
    email: '',
    password: '',
    confirmPassword: '',
  });

  const { name, email, password, confirmPassword } = values;

  const handleRegister = (e) => {
    e.preventDefault();

    isFormValid() &&
      dispatch(startRegisterWithEmailPasswordName({ email, password, name }));
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(uiSetErrorMessage(messageErrorName));
      return false;
    }

    if (!isEmail(email)) {
      dispatch(uiSetErrorMessage(messageErrorEmail));
      return false;
    }

    if (
      password.trim() !== confirmPassword.trim() ||
      (!isStrongPassword(password) && !isStrongPassword(confirmPassword))
    ) {
      dispatch(uiSetErrorMessage(messageErrorPassword));
      return false;
    }

    messageError && dispatch(uiRemoveErrorMessage());
    return true;
  };

  return (
    <>
      <h3 className="auth__title">{REGISTER}</h3>
      <form
        onSubmit={handleRegister}
        className="animate__animated animate__fadeIn animate_faster"
      >
        {messageError && (
          <div className="auth__alert-error">{messageError}</div>
        )}

        <input
          type="text"
          placeholder="Full Name"
          name="name"
          autoComplete="off"
          className="auth__input"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Confirm password"
          name="confirmPassword"
          className="auth__input"
          value={confirmPassword}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary btn-block mb-5">
          {REGISTER}
        </button>

        <Link to="/auth/login" className="link mt-5">
          {ALREADY_REGISTER}
        </Link>
      </form>
    </>
  );
};

export default RegisterScreen;
