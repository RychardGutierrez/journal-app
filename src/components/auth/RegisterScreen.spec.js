import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configure, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

import RegisterScreen from './RegisterScreen';

jest.mock('../../actions/auth', () => ({
  startRegisterWithEmailPasswordName: jest.fn(),
}));

configure({ adapter: new Adapter() });
const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('<RegisterScreen>', () => {
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  it('should render Register screen', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should register user', () => {
    wrapper.find('input[name="name"]').simulate('change', {
      target: { value: 'Lucas the duck', name: 'name' },
    });
    wrapper.find('input[name="email"]').simulate('change', {
      target: { value: 'lucas.duck@gmail.com', name: 'email' },
    });
    wrapper.find('input[name="password"]').simulate('change', {
      target: { value: '123Lucas@', name: 'password' },
    });
    wrapper.find('input[name="confirmPassword"]').simulate('change', {
      target: { value: '123Lucas@', name: 'confirmPassword' },
    });
    wrapper.find('form').prop('onSubmit')({ preventDefault: () => {} });

    expect(startRegisterWithEmailPasswordName).toHaveBeenLastCalledWith({
      email: 'lucas.duck@gmail.com',
      name: 'Lucas the duck',
      password: '123Lucas@',
    });
  });
});
