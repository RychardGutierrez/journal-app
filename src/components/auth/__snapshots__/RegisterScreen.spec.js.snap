// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen> should render Register screen 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate_faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Full Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link mt-5"
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>,
]
`;
